rsocket-core/src/main/java/io/rsocket/lease/LeaseRSocketRef.java

```diff
@@ -0,0 +1,106 @@
+package io.rsocket.lease;
+
+import io.rsocket.Availability;
+import io.rsocket.Closeable;
+import io.rsocket.RSocket;
+import java.nio.ByteBuffer;
+import java.util.Objects;
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import reactor.core.publisher.Mono;
+
+/** Provides means to grant lease to peer and check RSocket responder availability */
+public class LeaseRSocketRef implements Availability {
+  private static final CloseableAvailability closed = new ClosedAvailability();
+  private final LeaseGranter leaseGranter;
+  private volatile CloseableAvailability rsocket;
+  private volatile boolean isClosed;
+
+  LeaseRSocketRef(@Nonnull LeaseGranter leaseGranter, @Nonnull RSocket responderRSocket) {
+    Objects.requireNonNull(leaseGranter, "leaseGranter");
+    Objects.requireNonNull(responderRSocket, "responderRSocket");
+    this.leaseGranter = leaseGranter;
+    this.rsocket = wrap(responderRSocket);
+    this.rsocket
+        .onClose()
+        .doOnTerminate(
+            () -> {
+              this.isClosed = true;
+              this.rsocket = closed;
+            })
+        .subscribe();
+  }
+
+  public double availability() {
+    return rsocket.availability();
+  }
+
+  public boolean isClosed() {
+    return isClosed;
+  }
+
+  public void grantLease(int numberOfRequests, long timeToLive, @Nullable ByteBuffer metadata) {
```

Comment:
metadata on received lease is not exposed yet. Should it be ByteBuffer or whole Lease?